File: examples/src/main/java/com/edgedb/examples/CustomDeserializer.java
package com.edgedb.examples;

import com.edgedb.driver.EdgeDBClient;
import com.edgedb.annotations.EdgeDBDeserializer;
import com.edgedb.annotations.EdgeDBName;
import com.edgedb.annotations.EdgeDBType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletionStage;

public final class CustomDeserializer implements Example {
    private static final Logger logger = LoggerFactory.getLogger(CustomDeserializer.class);

    @EdgeDBType
    public static final class Person {
        private final String name;
        private final Long age;

        @EdgeDBDeserializer
        public Person(
                @EdgeDBName("name") String name,
                @EdgeDBName("age") Long age
        ) {
            this.name = name;
            this.age = age;

            logger.info("Custom deserializer called");
        }
    }

    @Override
    public CompletionStage<Void> run(EdgeDBClient client) {
        return client.execute("insert Person { name := 'Example', age := 123 } unless conflict on .name")
                .thenCompose(v -> client.queryRequiredSingle(Person.class, "select Person { name, age } filter .name = 'Example'"))
                .thenAccept(person -> logger.info("Person result: {person}"));
    }
}

File: examples/kotlin-examples/src/main/kotlin/com/gel/examples/CustomDeserializer.kt
package com.gel.examples

import com.gel.driver.GelClientPool
import com.gel.driver.annotations.GelDeserializer
import com.gel.driver.annotations.GelName
import com.gel.driver.annotations.GelType
import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory

class CustomDeserializer : Example {
    companion object {
        private val logger = LoggerFactory.getLogger(CustomDeserializer::class.java)!!
    }

    @GelType
    data class Person
    @GelDeserializer
    constructor (
            @GelName("name")
            val name: String,
            @GelName("age")
            val age: Long
    ) {
        init {
            logger.info("Custom deserializer called with: name: {}, age: {}", name, age)
        }
    }

    override suspend fun runAsync(clientPool: GelClientPool) {
        val person = clientPool.queryRequiredSingle(
                Person::class.java,
                """
                    insert Person { name := 'Example', age := 123 } unless conflict on .name;
                    select Person { name, age } filter .name = 'Example'
                """.trimIndent()
        ).await()

        logger.info("Got person: {}", person)
    }
}